File: Frameworks/SoSeedySwift/SoSeedySwift/VCR/VCR.m
//
// Copyright (C) 2018-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import "VCR.h"

static NSDictionary *requests = nil;
static VCR *_shared = nil;

@implementation VCR
{
    BOOL _record;
};

RCT_EXPORT_MODULE();

+(void) initialize
{
    if (!requests) {
        requests = [[NSMutableDictionary alloc] init];
    }
}

- (instancetype)init
{
    self = [super init];
    if (_shared == nil) {
        _shared = self;
    }
    _record = YES;
    return self;
}

+ (VCR *)shared
{
    return _shared;
}

- (NSURL *)testCaseToUrl:(NSString *)testCase
{
    NSString *fileName = [testCase stringByReplacingOccurrencesOfString:@"-" withString:@""];
    fileName = [fileName stringByReplacingOccurrencesOfString:@"[" withString:@""];
    fileName = [fileName stringByReplacingOccurrencesOfString:@"]" withString:@""];
    fileName = [fileName stringByReplacingOccurrencesOfString:@" " withString:@"_"];
    
    // this URL will need to be replaced when we decide where we will be uploading the cassette files
    NSString *urlString = [NSString stringWithFormat:@"http://localhost:9000/cassettes/%@.json", fileName];
    NSURL *url = [NSURL URLWithString:urlString];
    return url;
}

- (void)loadCassette:(NSString *)testCase completionHandler:(void (^)(NSError *error))completionHandler
{
    if (_record) {
        requests = [[NSMutableDictionary alloc] init];
        return completionHandler(nil);
    }
    NSData *jsonData = [[NSData alloc] initWithContentsOfURL:[self testCaseToUrl:testCase]];
    NSError *error;
    requests = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingMutableContainers error:&error];
    completionHandler(error);
}

- (void)recordCassette:(NSString *)testCase completionHandler:(void (^)(NSError *error))completionHandler
{
    // This line is only here to prevent from actually uploading the cassette files
    // as we don't have anywhere to upload them yet
    return completionHandler(nil);
    
    if (!_record) {
        return completionHandler(nil);
    }
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:requests options:nil error:&error];
    
    if (error == nil) {
        [self uploadCassetteFile:testCase jsonData:jsonData completionHandler:completionHandler];
    } else {
        [completionHandler error];
    }
}

- (void)uploadCassetteFile:(NSString *)testCase jsonData:(NSData *)jsonData completionHandler:(void (^)(NSError *error))completionHandler
{
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[self testCaseToUrl:testCase]];
    
    [request setHTTPMethod:@"PUT"];
    
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    
    NSMutableData *body = [NSMutableData data];
    [body appendData:jsonData];
    
    [request setHTTPBody:body];
    
    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        completionHandler(error);
    }];
    [task resume];
}

- (dispatch_queue_t)methodQueue {
    return dispatch_get_main_queue();
}

- (NSString *)stringFrom:(NSDictionary *)dictionary
{
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:requests options:nil error:&error];
    if (error != nil) {
        return nil;
    }
    NSString *key = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    return key;
}

- (void)recordResponse:(NSString *)value for:(NSString *)key
{
    if (!_record) {
        return;
    }
    NSLog(key);
    NSLog(value);
    [requests setValue:requests[key] ? requests[key] : [[NSMutableArray alloc] init] forKey:key];
    [requests[key] addObject:value];
}

RCT_REMAP_METHOD(recordRequest, recordRequest:(NSString *) requestConfig responseConfig:(NSString *)responseConfig resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    if (!_record) {
        return resolve(nil);
    }
    [requests setValue:(requests[requestConfig] ? requests[requestConfig] : [[NSMutableArray alloc] init]) forKey:requestConfig];
    [requests[requestConfig] addObject:responseConfig];
    resolve(nil);
    
}

- (NSString *)responseFor:(NSString *)key {
    if (_record) {
        return nil;
    }

    if (requests[key]) {
        NSString *response = [requests[key] firstObject];
        [requests[key] removeObjectAtIndex:0];
        return response;
    }
    return nil;
}

RCT_REMAP_METHOD(responseForRequest, responseForRequest:(NSString *) requestConfig resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    if (_record) {
        return resolve(nil);
    }
    if (requests[requestConfig]) {
        NSString *response = [requests[requestConfig] firstObject];
        resolve(response);
        [requests[requestConfig] removeObjectAtIndex:0];
    } else {
        resolve(nil);
    }
}

@end

File: Frameworks/SoSeedySwift/SoSeedySwift/VCR.swift
//
// Copyright (C) 2018-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import Foundation

var requests: [String : [String]] = [:]

public class VCR {
    public var record = true
    public static let shared = VCR()
    
    init() {}
    
    public func testCaseToURL(_ testCase: String) -> URL? {
        var fileName = testCase.replacingOccurrences(of: "-", with: "")
        fileName = fileName.replacingOccurrences(of: "[", with: "")
        fileName = fileName.replacingOccurrences(of: "]", with: "")
        fileName = fileName.replacingOccurrences(of: " ", with: "")
        
        let urlString = "http://localhost:9000/cassettes/\(fileName).json"
        return URL(string: urlString)
    }
    
    public func loadCassette(testCase: String, completionHandler: (Error?) -> Void) {
        guard record == false, let url = testCaseToURL(testCase) else {
            requests = [:]
            completionHandler(nil)
            return
        }
        do {
            let jsonData = try Data(contentsOf: url)
            requests = try JSONSerialization.jsonObject(with: jsonData, options: .mutableContainers) as! [String : [String]]
            completionHandler(nil)
        } catch let error {
            completionHandler(error)
        }
    }
    
    public func recordCassette(testCase: String, completionHandler:@escaping (Error?) -> Void) {
        if !record {
            completionHandler(nil)
            return
        }
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: requests)
            uploadCassetteFile(testCase: testCase, jsonData: jsonData, completionHandler:completionHandler)
        } catch let error {
            completionHandler(error)
        }
    }
    
    public func uploadCassetteFile(testCase: String, jsonData: Data, completionHandler: @escaping (Error?) -> Void) {
        guard let testCaseURL = testCaseToURL(testCase) else {
            completionHandler(nil)
            return
        }
        var request = URLRequest(url: testCaseURL)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData
        
        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            completionHandler(error)
        }
        task.resume()
    }
    
    public func recordResponse(_ value: String, for key: String) {
        if !record {
            return
        }
        requests[key] = requests[key] ?? []
        requests[key]?.append(value)
    }
    
    public func response(for key:String) -> String? {
        if record {
            return nil
        }
        
        guard var value = requests[key], let response = value.first else {
            return nil
        }
        value.remove(at: 0)
        requests[key] = value
        return response
    }
}
